packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js

```diff
@@ -162,312 +105,76 @@ RocketChat.callbacks.add('afterSaveMessage', function(message, room, userId) {
-				});
-			}
-
-			userIdsToNotify = _.pluck(usersOfDesktopMentions, '_id');
-		}
-
-		if (mentionIds.length + settings.alwaysNotifyMobileUsers.length > 0) {
-			let mobileMentionIds = _.union(mentionIds, settings.alwaysNotifyMobileUsers);
-			mobileMentionIds = _.difference(mobileMentionIds, settings.dontNotifyMobileUsers);
-
-			const usersOfMobileMentionsQuery = {
-				_id: {
-					$in: mobileMentionIds
-				}
-			};
-
-			if (alwaysNotifyMobileBoolean !== true) {
-				usersOfMobileMentionsQuery.statusConnection = { $ne: 'online' };
-			}
-
-			let usersOfMobileMentions = RocketChat.models.Users.find(usersOfMobileMentionsQuery, {
-				fields: {
-					_id: 1,
-					username: 1,
-					statusConnection: 1,
-					active: 1
-				}
-			}).fetch();
-
-			mentions.push(...usersOfMobileMentions);
-			if (room.t !== 'c') {
-				usersOfMobileMentions = _.reject(usersOfMobileMentions, usersOfMentionItem => !room.usernames.includes(usersOfMentionItem.username));
-			}
+		const mentionsFilter = { $in: ['all', 'mentions'] };
+		const excludesNothingFilter = { $ne: 'nothing' };
+
+		subscriptions = RocketChat.models.Subscriptions.findNotificationPreferencesByRoom({
+			roomId: room._id,
+			desktopFilter: RocketChat.settings.get('Accounts_Default_User_Preferences_desktopNotifications') === 'nothing' ? mentionsFilter : excludesNothingFilter,
+			emailFilter: RocketChat.settings.get('Accounts_Default_User_Preferences_emailNotificationMode') === 'disabled' ? mentionsFilter : excludesNothingFilter,
+			mobileFilter: RocketChat.settings.get('Accounts_Default_User_Preferences_mobileNotifications') === 'nothing' ? mentionsFilter : excludesNothingFilter
+		});
+	}
 
-			userIdsToPushNotify = usersOfMobileMentions.map(userMobile => {
-				pushUsernames[userMobile._id] = userMobile.username;
-				return userMobile._id;
-			});
-		}
-
-		if (mentionIds.length + settings.alwaysNotifyAudioUsers.length > 0) {
-			let audioMentionIds = _.union(mentionIds, settings.alwaysNotifyAudioUsers);
-			audioMentionIds = _.difference(audioMentionIds, userIdsToNotify);
-
-			let usersOfAudioMentions = RocketChat.models.Users.find({ _id: { $in: audioMentionIds }, statusConnection: {
-				$ne:'offline'
-			} }, {
-				fields: {
-					_id: 1,
-					username: 1,
-					active: 1
-				}
-			}).fetch();
-			mentions.push(...usersOfAudioMentions);
-			if (room.t !== 'c') {
-				usersOfAudioMentions = _.reject(usersOfAudioMentions, (usersOfMentionItem) => {
-					return room.usernames.indexOf(usersOfMentionItem.username) === -1;
+	const mentionIds = (message.mentions || []).map(({_id}) => _id);
+	const hasMentionToAll = mentionIds.includes('all');
+	const hasMentionToHere = mentionIds.includes('here');
+
+	subscriptions.forEach((subscription) => sendNotification({
+		subscription,
+		sender,
+		hasMentionToAll,
+		hasMentionToHere,
+		message,
+		room,
+		mentionIds,
+		disableAllMessageNotifications
+	}));
+
+	if (room.t === 'c') {
```

Comment:
on public channels, if a mentioned user is not member of the channel yet, he will first `join` the channel and then be notified based on his preferences.